src/client.ts

```diff
@@ -0,0 +1,864 @@
+
+type OrderItem = Chan["id"] | Network["uuid"];
+type Order = OrderItem[];
+
+const events = [
+	"away",
+	"cap",
+	"connection",
+	"unhandled",
+	"ctcp",
+	"chghost",
+	"error",
+	"help",
+	"info",
+	"invite",
+	"join",
+	"kick",
+	"list",
+	"mode",
+	"modelist",
+	"motd",
+	"message",
+	"names",
+	"nick",
+	"part",
+	"quit",
+	"sasl",
+	"topic",
+	"welcome",
+	"whois",
+];
+
+type ClientPushSubscription = {
+	endpoint: string;
+	keys: {
+		p256dh: string;
+		auth: string;
+	};
+};
+
+export type UserConfig = {
+	log: boolean;
+	password: string;
+	sessions: {
+		[token: string]: {
+			lastUse: number;
+			ip: string;
+			agent: string;
+			pushSubscription?: ClientPushSubscription;
+		};
+	};
+	clientSettings: {
+		[key: string]: any;
+	};
+	browser?: {
+		language?: string;
+		ip?: string;
+		hostname?: string;
+		isSecure?: boolean;
+	};
+};
+
+export type Mention = {
+	chanId: number;
+	msgId: number;
+	type: MessageType;
+	time: Date;
+	text: string;
+	from: UserInMessage;
+};
+
+class Client {
+	awayMessage!: string;
+	lastActiveChannel!: number;
+	attachedClients!: {
+		[socketId: string]: {token: string; openChannel: number};
+	};
+	config!: UserConfig & {
+		networks?: Network[];
+	};
+	id!: number;
+	idMsg!: number;
+	idChan!: number;
+	name!: string;
+	networks!: Network[];
+	mentions!: Mention[];
+	manager!: ClientManager;
+	messageStorage!: MessageStorage[];
+	highlightRegex!: RegExp | null;
+	highlightExceptionRegex!: RegExp | null;
+	messageProvider?: SqliteMessageStorage;
+
+	fileHash!: string;
+
+	constructor(manager: ClientManager, name?: string, config = {} as UserConfig) {
+		_.merge(this, {
+			awayMessage: "",
+			lastActiveChannel: -1,
+			attachedClients: {},
+			config: config,
+			id: uuidv4(),
+			idChan: 1,
+			idMsg: 1,
+			name: name,
+			networks: [],
+			mentions: [],
+			manager: manager,
+			messageStorage: [],
+			highlightRegex: null,
+			highlightExceptionRegex: null,
+			messageProvider: undefined,
+		});
+
+		const client = this;
+
+		client.config.log = Boolean(client.config.log);
+		client.config.password = String(client.config.password);
+
+		if (!Config.values.public && client.config.log) {
+			if (Config.values.messageStorage.includes("sqlite")) {
+				client.messageProvider = new SqliteMessageStorage(client);
+				client.messageStorage.push(client.messageProvider);
+			}
+
+			if (Config.values.messageStorage.includes("text")) {
+				client.messageStorage.push(new TextFileMessageStorage(client));
+			}
+
+			for (const messageStorage of client.messageStorage) {
+				messageStorage.enable();
+			}
+		}
+
+		if (!_.isPlainObject(client.config.sessions)) {
+			client.config.sessions = {};
+		}
+
+		if (!_.isPlainObject(client.config.clientSettings)) {
+			client.config.clientSettings = {};
+		}
+
+		if (!_.isPlainObject(client.config.browser)) {
+			client.config.browser = {};
+		}
+
+		if (client.config.clientSettings.awayMessage) {
```

Comment:
fwiw backwards compat for `client.config.awayMessage` was removed

Reply:
added to #4578